src/openms/source/ANALYSIS/ID/SimpleSearchEngineAlgorithm.cpp

```diff
@@ -0,0 +1,654 @@
+
+          multimap_mass_2_scan_index.insert(make_pair(precursor_mass, scan_index));
+        }
+      }
+    }
+
+    // create spectrum generator
+    TheoreticalSpectrumGenerator spectrum_generator;
+    Param param(spectrum_generator.getParameters());
+    param.setValue("add_first_prefix_ion", "true");
+    param.setValue("add_metainfo", "true");
+    spectrum_generator.setParameters(param);
+
+    // preallocate storage for PSMs
+    vector<vector<AnnotatedHit_> > annotated_hits(spectra.size(), vector<AnnotatedHit_>());
+    for (auto & a : annotated_hits) { a.reserve(2 * report_top_hits_); }
+
+#ifdef _OPENMP
+    // we want to do locking at the spectrum level so we get good parallelisation 
+    vector<omp_lock_t> annotated_hits_lock(annotated_hits.size());
+    for (size_t i = 0; i != annotated_hits_lock.size(); i++) { omp_init_lock(&(annotated_hits_lock[i])); }
+#endif
+
+    startProgress(0, 1, "Load database from FASTA file...");
+    FASTAFile fastaFile;
+    vector<FASTAFile::FASTAEntry> fasta_db;
+    fastaFile.load(in_db, fasta_db);
+    endProgress();
+
+    ProteaseDigestion digestor;
+    digestor.setEnzyme(enzyme_);
+    digestor.setMissedCleavages(peptide_missed_cleavages_);
+
+    startProgress(0, (Size)(fasta_db.end() - fasta_db.begin()), "Scoring peptide models against spectra...");
+
+    // lookup for processed peptides. must be defined outside of omp section and synchronized
+    set<StringView> processed_petides;
+
+    Size count_proteins(0), count_peptides(0);
+
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static)
+#endif
+      for (SignedSize fasta_index = 0; fasta_index < (SignedSize)fasta_db.size(); ++fasta_index)
+      {
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
+      ++count_proteins;
+
+      IF_MASTERTHREAD
+      {
+        setProgress(count_proteins);
+      }
+
+      vector<StringView> current_digest;
+      digestor.digestUnmodified(fasta_db[fasta_index].sequence, current_digest, peptide_min_size_, peptide_max_size_);
+
+      for (auto const & c : current_digest)
+      { 
+        const String current_peptide = c.getString();
+        if (current_peptide.find_first_of("XBZ") != std::string::npos) { continue; }
+
+        // if a peptide motif is provided skip all peptides without match
+        if (!peptide_motif_.empty() && !boost::regex_match(current_peptide, peptide_motif_regex)) { continue; }          
+      
+        bool already_processed = false;
+#ifdef _OPENMP
+#pragma omp critical (processed_peptides_access)
+#endif
+        {
+          // peptide (and all modified variants) already processed so skip it
+          if (processed_petides.find(c) != processed_petides.end())
+          {
+            already_processed = true;
+          }
+        }
+
+        // skip peptides that have already been processed
+        if (already_processed) { continue; }
+
+#ifdef _OPENMP
+#pragma omp critical (processed_peptides_access)
+#endif
+        {
+          processed_petides.insert(c);
```

Comment:
I think you would want to have that in the same `omp critical` section, otherwise you could have a race condition: two identical peptides start at the top, both find that they are not in `processed_petides` and then they both progress to here, which is exactly what you want to avoid. I think you need to have

```

#ifdef _OPENMP
#pragma omp critical (processed_peptides_access)
#endif
        {
          // peptide (and all modified variants) already processed so skip it
          if (processed_petides.find(c) != processed_petides.end())
          {
            already_processed = true;
          }
       
          // skip peptides that have already been processed
          if (!already_processed) processed_petides.insert(c);
        }
        if (already_processed) { continue; }
```

this will prevent the race condition

Reply:
you are right. this could lead to the same peptide being scored twice. otherwise I think there should be no harm done. I will change it

Reply:
I changed it to:
```
#pragma omp critical (processed_peptides_access)
        { 
          // peptide (and all modified variants) already processed so skip it
          if (processed_petides.find(c) != processed_petides.end())
          { 
            already_processed = true;
          }
          else
          { 
            processed_petides.insert(c);
          }
        }             
                      
        // skip peptides that have already been processed
        if (already_processed) { continue; }
```
and wondered if putting the continue statement in the critical section is allowed.
@cbielow or @hroest I did not find this in the specification. Do you know if that is allowed?